src/process_quantiles.R

```diff
@@ -0,0 +1,184 @@
+  last_doy <- 365 # switch back to 366 once leap days are implemented
+  if(begin_doy <= 0) {
+    # Need to look backward into last days of the year
+    seq_doy <- c((last_doy + begin_doy):last_doy, 1:end_doy)
+  } else if (end_doy > last_doy) {
+    # Need to look forward into first days of the year
+    seq_doy <- c(begin_doy:last_doy, 1:(end_doy - last_doy))
+  } else {
+    seq_doy <- begin_doy:end_doy
+  }
+  
+  return(seq_doy)
+}
+extract_doy_cols_to_vec <- function(df, doy_seq) {
+  ############## WORKING ON THIS
+  # Do I need to convert to df?
+  df <- as.data.frame(df)
+  doy_seq_colnames <- sprintf("%03d", doy_seq)
+  target_df <- df[, doy_seq_colnames]
+  ############## 
+  target_vector <- unlist(target_df, use.names = FALSE)
+  return(target_vector)
+}
+get_quantile_df <- function(target_doy_values, target_doy_seq, 
+                            target_doy, probs = c(0.10, 0.25, 0.75, 0.90)) {
+  target_doy_quantiles <- quantile(target_doy_values, 
+                                   probs = probs, 
+                                   type = 6,
+                                   na.rm = TRUE)
+  target_doy_quantile_df <- cbind(DOY = target_doy, t(target_doy_quantiles))
+  return(target_doy_quantile_df)
+}
+########## ////// ########## 
+
+# Identify task id from yeti environment & convert to HRU ids ----
+# task_id <- as.numeric(Sys.getenv('SLURM_ARRAY_TASK_ID', 'NA'))
+task_id <- 2 # Will uncomment once setup with slurm
+hru_seq <- task_id_to_hru_seq(task_id)
+
+message(sprintf("Started this task at %s", Sys.time()))
+
+vars <- c(
+  "soil_moist_tot",
+  "pkwater_equiv",
+  "hru_intcpstor", 
+  "hru_impervstor",
+  "gwres_stor", 
+  "dprst_stor_hru"
+)
+
+dt_long <- c()
+for(var in vars) {
+  
+  # Read in the two datasets
+  ncdf_fn <- sprintf("historical_%s_out.nc", var)
+  var_df <- read_ncdf_data(ncdf_fn, var, hru_seq)
+  dt_long <- rbind(dt_long, var_df)
+}
+
+#######################
+# Find data.table equivalent to "separate"
+# dt_long_sep <- tidyr::separate(dt_long, year_doy, c("year", "doy"), "_")
+# Doesn't seem that much faster ...
+dt_long[, c("year", "doy") := tstrsplit(year_doy, "_", fixed=TRUE)]
+dt_long[, year_doy := NULL]
+
+#######################
+
+# `fill = NA` to account for missing dates at beginning 
+#   of 1980 and end of 2019
+# This spreads + sums the values! Takes ~ 10 sec
+dt_wide <- dcast(dt_long, hruid+year ~ doy, fun=sum, fill=NA)
+dt_wide[, year := NULL]
+
+all_hru_quantiles_list <- lapply(hru_seq, function(hruid_i) {
+  
+  hru_data <- dt_wide[hruid == hruid_i,]
+  hru_data[, hruid := NULL]
+  
+  doy_quantile_list <- lapply(1:365, function(target_doy, df) {
+    target_doy_seq <- get_doy_sequence(target_doy)
+    target_doy_values <- extract_doy_cols_to_vec(df, target_doy_seq)
+    target_doy_quantile_df <- get_quantile_df(target_doy_values, target_doy_seq, target_doy)
+    return(target_doy_quantile_df)
+  }, df = hru_data)
+  
+  hru_quantiles_df <- do.call("rbind", doy_quantile_list)
```

Comment:
I thought for sure that pre-allocating a matrix to hold this info would speed things up...but as near as I can tell, this do.call(rbind) really is faster. And more readable, too - nice!